File: src/main/resources/static/js/jaffolding/ui/DataTable.js
/**
 * A data table component for displaying tabular data.
 * JavaScript implementation that mirrors the Java API.
 */
(function() {
  class DataTable extends jaffolding.Component {
    /**
     * Creates a new DataTable instance.
     * @param {string[]} [columnNames] - The column names
     */
    constructor(columnNames = []) {
      super('div');
      this.columnNames = [...columnNames];
      this.data = [];
      this.dataState = new jaffolding.State([]);
      this.selectable = true;
      this.selectedRow = -1;
      this.selectionListener = null;
      this.rowClickListener = null;
      this.sortColumn = null;
      this.sortAscending = true;
      this.columnTypes = {};
      this.filters = {};
      
      this.initializeStyles();
      
      // Subscribe to data state changes
      this.dataState.subscribe(newData => {
        this.data = newData;
        this.refreshTable();
      });
    }

    /**
     * Initializes the table styles.
     * @private
     */
    initializeStyles() {
      this.setStyle('width', '100%')
          .setStyle('overflow', 'auto')
          .setStyle('border', '1px solid #e0e0e0')
          .setStyle('border-radius', '4px')
          .setStyle('background-color', 'white');
    }

    /**
     * Sets the column names.
     * @param {string[]} columnNames - The column names
     * @returns {DataTable} - The data table instance for chaining
     */
    setColumnNames(columnNames) {
      this.columnNames = [...columnNames];
      this.refreshTable();
      return this;
    }

    /**
     * Sets the column type.
     * @param {string} columnName - The column name
     * @param {string} type - The column type ('string', 'number', 'date')
     * @returns {DataTable} - The data table instance for chaining
     */
    setColumnType(columnName, type) {
      this.columnTypes[columnName] = type;
      return this;
    }

    /**
     * Sets the table data.
     * @param {Object[]} data - The table data
     * @returns {DataTable} - The data table instance for chaining
     */
    setData(data) {
      const newData = data.map(row => ({...row}));
      this.dataState.set(newData);
      return this;
    }

    /**
     * Adds a row to the table.
     * @param {Object} row - The row data
     * @returns {DataTable} - The data table instance for chaining
     */
    addRow(row) {
      const newData = [...this.data, {...row}];
      this.dataState.set(newData);
      return this;
    }

    /**
     * Removes a row from the table.
     * @param {number} index - The row index
     * @returns {DataTable} - The data table instance for chaining
     */
    removeRow(index) {
      if (index >= 0 && index < this.data.length) {
        const newData = [...this.data];
        newData.splice(index, 1);
        this.dataState.set(newData);
        
        if (this.selectedRow >= this.data.length) {
          this.selectedRow = this.data.length - 1;
        }
      }
      return this;
    }

    /**
     * Updates a row in the table.
     * @param {number} index - The row index
     * @param {Object} row - The new row data
     * @returns {DataTable} - The data table instance for chaining
     */
    updateRow(index, row) {
      if (index >= 0 && index < this.data.length) {
        const newData = [...this.data];
        newData[index] = {...row};
        this.dataState.set(newData);
      }
      return this;
    }

    /**
     * Clears the table data.
     * @returns {DataTable} - The data table instance for chaining
     */
    clearData() {
      this.dataState.set([]);
      return this;
    }

    /**
     * Sets whether the table is selectable.
     * @param {boolean} selectable - Whether the table is selectable
     * @returns {DataTable} - The data table instance for chaining
     */
    setSelectable(selectable) {
      this.selectable = selectable;
      this.refreshTable();
      return this;
    }

    /**
     * Gets the selected row index.
     * @returns {number} - The selected row index
     */
    getSelectedRow() {
      return this.selectedRow;
    }

    /**
     * Gets the selected row data.
     * @returns {Object|null} - The selected row data or null if no row is selected
     */
    getSelectedRowData() {
      if (this.selectedRow >= 0 && this.selectedRow < this.data.length) {
        return this.data[this.selectedRow];
      }
      return null;
    }

    /**
     * Sets the selected row.
     * @param {number} row - The row index
     * @returns {DataTable} - The data table instance for chaining
     */
    setSelectedRow(row) {
      if (row >= -1 && row < this.data.length) {
        this.selectedRow = row;
        this.updateSelection();
      }
      return this;
    }

    /**
     * Sets the row selection listener.
     * @param {Function} listener - The selection listener function
     * @returns {DataTable} - The data table instance for chaining
     */
    setOnRowSelect(listener) {
      this.selectionListener = listener;
      return this;
    }

    /**
     * Sets the row click listener.
     * @param {Function} listener - The click listener function
     * @returns {DataTable} - The data table instance for chaining
     */
    setOnRowClick(listener) {
      this.rowClickListener = listener;
      return this;
    }

    /**
     * Sorts the table by a column.
     * @param {string} columnName - The column name
     * @param {boolean} ascending - Whether to sort in ascending order
     * @returns {DataTable} - The data table instance for chaining
     */
    sortBy(columnName, ascending) {
      this.sortColumn = columnName;
      this.sortAscending = ascending;
      
      if (this.data.length > 0 && columnName) {
        const newData = [...this.data];
        
        newData.sort((a, b) => {
          const valA = a[columnName];
          const valB = b[columnName];
          
          if (valA === undefined && valB === undefined) return 0;
          if (valA === undefined) return ascending ? -1 : 1;
          if (valB === undefined) return ascending ? 1 : -1;
          
          const type = this.columnTypes[columnName] || 'string';
          let result = 0;
          
          switch (type) {
            case 'number':
              const numA = typeof valA === 'number' ? valA : parseFloat(valA);
              const numB = typeof valB === 'number' ? valB : parseFloat(valB);
              result = numA - numB;
              break;
            case 'date':
              const dateA = new Date(valA);
              const dateB = new Date(valB);
              result = dateA - dateB;
              break;
            default:
              result = String(valA).localeCompare(String(valB));
          }
          
          return ascending ? result : -result;
        });
        
        this.dataState.set(newData);
      }
      
      return this;
    }

    /**
     * Filters the table by a column.
     * @param {string} columnName - The column name
     * @param {string} value - The filter value
     * @returns {DataTable} - The data table instance for chaining
     */
    filter(columnName, value) {
      if (!value || value.trim() === '') {
        delete this.filters[columnName];
      } else {
        this.filters[columnName] = value.toLowerCase();
      }
      
      this.applyFilters();
      return this;
    }

    /**
     * Clears all filters.
     * @returns {DataTable} - The data table instance for chaining
     */
    clearFilters() {
      this.filters = {};
      this.dataState.set(this.data);
      return this;
    }

    /**
     * Applies all filters to the data.
     * @private
     */
    applyFilters() {
      if (Object.keys(this.filters).length === 0) {
        return;
      }
      
      const filteredData = this.data.filter(row => {
        for (const [columnName, filterValue] of Object.entries(this.filters)) {
          if (row[columnName] === undefined || row[columnName] === null) {
            return false;
          }
          
          const cellValue = String(row[columnName]).toLowerCase();
          if (!cellValue.includes(filterValue)) {
            return false;
          }
        }
        
        return true;
      });
      
      this.dataState.set(filteredData);
    }

    /**
     * Loads data from an API endpoint.
     * @param {string} url - The API endpoint URL
     * @returns {DataTable} - The data table instance for chaining
     */
    loadFromApi(url) {
      fetch(url)
        .then(response => response.json())
        .then(data => {
          this.dataState.set(data);
        })
        .catch(error => {
          console.error('Error loading data from API:', error);
        });
      
      return this;
    }

    /**
     * Refreshes the table.
     * @private
     */
    refreshTable() {
      if (!this.element) {
        return;
      }
      
      this.element.innerHTML = '';
      
      // Create table element
      const table = document.createElement('table');
      table.style.width = '100%';
      table.style.borderCollapse = 'collapse';
      table.style.fontSize = '14px';
      
      // Create header
      if (this.columnNames.length > 0) {
        const thead = document.createElement('thead');
        const headerRow = document.createElement('tr');
        
        for (const columnName of this.columnNames) {
          const th = document.createElement('th');
          th.textContent = columnName;
          th.style.padding = '12px 10px';
          th.style.textAlign = 'left';
          th.style.borderBottom = '2px solid #ddd';
          th.style.fontWeight = '600';
          th.style.backgroundColor = '#f5f5f5';
          th.style.cursor = 'pointer';
          
          // Add sort indicator if this column is sorted
          if (columnName === this.sortColumn) {
            th.textContent = `${columnName} ${this.sortAscending ? '▲' : '▼'}`;
          }
          
          // Add sort functionality
          th.addEventListener('click', () => {
            const asc = columnName !== this.sortColumn || !this.sortAscending;
            this.sortBy(columnName, asc);
          });
          
          headerRow.appendChild(th);
        }
        
        thead.appendChild(headerRow);
        table.appendChild(thead);
      }
      
      // Create filter row
      if (this.columnNames.length > 0 && Object.keys(this.filters).length > 0) {
        const filterRow = document.createElement('tr');
        filterRow.style.backgroundColor = '#f9f9f9';
        
        for (const columnName of this.columnNames) {
          const td = document.createElement('td');
          td.style.padding = '8px 10px';
          
          if (this.filters[columnName]) {
            const input = document.createElement('input');
            input.type = 'text';
            input.placeholder = 'Filter...';
            input.value = this.filters[columnName];
            input.style.width = '100%';
            input.style.padding = '4px';
            input.style.border = '1px solid #ddd';
            input.style.borderRadius = '3px';
            
            input.addEventListener('input', () => {
              this.filter(columnName, input.value);
            });
            
            td.appendChild(input);
          }
          
          filterRow.appendChild(td);
        }
        
        table.appendChild(filterRow);
      }
      
      // Create body
      const tbody = document.createElement('tbody');
      
      for (let i = 0; i < this.data.length; i++) {
        const rowData = this.data[i];
        const row = document.createElement('tr');
        
        if (this.selectable) {
          row.style.cursor = 'pointer';
          
          row.addEventListener('click', () => {
            this.selectedRow = i;
            this.updateSelection();
            
            if (this.selectionListener) {
              this.selectionListener(i);
            }
            
            if (this.rowClickListener) {
              this.rowClickListener(rowData);
            }
          });
          
          if (i === this.selectedRow) {
            row.style.backgroundColor = '#e8f0fe';
          }
        }
        
        // Add hover effect
        row.addEventListener('mouseover', () => {
          if (i !== this.selectedRow) {
            row.style.backgroundColor = '#f5f5f5';
          }
        });
        
        row.addEventListener('mouseout', () => {
          if (i !== this.selectedRow) {
            row.style.backgroundColor = '';
          }
        });
        
        // Add cells
        for (const columnName of this.columnNames) {
          const td = document.createElement('td');
          td.style.padding = '10px';
          td.style.borderBottom = '1px solid #ddd';
          
          const cellValue = rowData[columnName];
          td.textContent = cellValue !== undefined ? String(cellValue) : '';
          
          row.appendChild(td);
        }
        
        tbody.appendChild(row);
      }
      
      table.appendChild(tbody);
      this.element.appendChild(table);
      
      // Add empty state message if no data
      if (this.data.length === 0) {
        const emptyState = document.createElement('div');
        emptyState.textContent = 'No data available';
        emptyState.style.padding = '20px';
        emptyState.style.textAlign = 'center';
        emptyState.style.color = '#888';
        
        this.element.appendChild(emptyState);
      }
    }

    /**
     * Updates the selection.
     * @private
     */
    updateSelection() {
      if (!this.element) {
        return;
      }
      
      const table = this.element.querySelector('table');
      if (!table) return;
      
      const tbody = table.querySelector('tbody');
      if (!tbody) return;
      
      for (let i = 0; i < tbody.childNodes.length; i++) {
        const row = tbody.childNodes[i];
        
        if (i === this.selectedRow) {
          row.style.backgroundColor = '#e8f0fe';
        } else {
          row.style.backgroundColor = '';
        }
      }
    }

    /**
     * Gets the table data.
     * @returns {Object[]} - The table data
     */
    getData() {
      return [...this.data];
    }

    /**
     * Gets the data state.
     * @returns {State} - The data state
     */
    getDataState() {
      return this.dataState;
    }
  }

  // Export the DataTable class
  window.jaffolding = window.jaffolding || {};
  window.jaffolding.DataTable = DataTable;
})();

File: src/main/java/com/danielremsburg/jaffolding/ui/DataTable.java
package com.danielremsburg.jaffolding.ui;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

import org.teavm.jso.JSBody;
import org.teavm.jso.JSObject;
import org.teavm.jso.ajax.XMLHttpRequest;
import org.teavm.jso.dom.events.Event;
import org.teavm.jso.dom.html.HTMLDocument;
import org.teavm.jso.dom.html.HTMLElement;
import org.teavm.jso.json.JSON;

import com.danielremsburg.jaffolding.Component;
import com.danielremsburg.jaffolding.State;
import com.danielremsburg.jaffolding.bridge.HTMLTableElements.HTMLTableElement;
import com.danielremsburg.jaffolding.bridge.HTMLTableElements.HTMLTableRowElement;

/**
 * An enhanced data table component with sorting, filtering, and data binding.
 */
public class DataTable extends Component {
    private List<String> columnNames = new ArrayList<>();
    private List<Map<String, Object>> data = new ArrayList<>();
    private State<List<Map<String, Object>>> dataState;
    private boolean selectable = true;
    private int selectedRow = -1;
    private Consumer<Integer> selectionListener;
    private Consumer<Map<String, Object>> rowClickListener;
    private String sortColumn = null;
    private boolean sortAscending = true;
    private Map<String, String> columnTypes = new HashMap<>();
    private Map<String, String> filters = new HashMap<>();
    
    public DataTable() {
        super("div");
        this.dataState = new State<>(new ArrayList<>());
        initializeStyles();
    }
    
    public DataTable(List<String> columnNames) {
        super("div");
        this.columnNames = new ArrayList<>(columnNames);
        this.dataState = new State<>(new ArrayList<>());
        initializeStyles();
    }
    
    private void initializeStyles() {
        setStyle("width", "100%")
            .setStyle("overflow", "auto")
            .setStyle("border", "1px solid #e0e0e0")
            .setStyle("border-radius", "4px")
            .setStyle("background-color", "white");
        
        // Subscribe to data state changes
        dataState.subscribe(newData -> {
            this.data = newData;
            refreshTable();
        });
    }
    
    public DataTable setColumnNames(List<String> columnNames) {
        this.columnNames = new ArrayList<>(columnNames);
        refreshTable();
        return this;
    }
    
    public DataTable setColumnType(String columnName, String type) {
        columnTypes.put(columnName, type);
        return this;
    }
    
    public DataTable setData(List<Map<String, Object>> data) {
        List<Map<String, Object>> newData = new ArrayList<>();
        for (Map<String, Object> row : data) {
            newData.add(new HashMap<>(row));
        }
        dataState.set(newData);
        return this;
    }
    
    public DataTable addRow(Map<String, Object> row) {
        List<Map<String, Object>> newData = new ArrayList<>(data);
        newData.add(new HashMap<>(row));
        dataState.set(newData);
        return this;
    }
    
    public DataTable removeRow(int index) {
        if (index >= 0 && index < data.size()) {
            List<Map<String, Object>> newData = new ArrayList<>(data);
            newData.remove(index);
            dataState.set(newData);
        }
        return this;
    }
    
    public DataTable updateRow(int index, Map<String, Object> row) {
        if (index >= 0 && index < data.size()) {
            List<Map<String, Object>> newData = new ArrayList<>(data);
            newData.set(index, new HashMap<>(row));
            dataState.set(newData);
        }
        return this;
    }
    
    public DataTable clearData() {
        dataState.set(new ArrayList<>());
        return this;
    }
    
    public DataTable setSelectable(boolean selectable) {
        this.selectable = selectable;
        refreshTable();
        return this;
    }
    
    public int getSelectedRow() {
        return selectedRow;
    }
    
    public Map<String, Object> getSelectedRowData() {
        if (selectedRow >= 0 && selectedRow < data.size()) {
            return data.get(selectedRow);
        }
        return null;
    }
    
    public DataTable setSelectedRow(int row) {
        if (row >= -1 && row < data.size()) {
            selectedRow = row;
            updateSelection();
        }
        return this;
    }
    
    public DataTable setOnRowSelect(Consumer<Integer> listener) {
        this.selectionListener = listener;
        return this;
    }
    
    public DataTable setOnRowClick(Consumer<Map<String, Object>> listener) {
        this.rowClickListener = listener;
        return this;
    }
    
    public DataTable sortBy(String columnName, boolean ascending) {
        this.sortColumn = columnName;
        this.sortAscending = ascending;
        
        if (data.size() > 0 && columnName != null) {
            List<Map<String, Object>> newData = new ArrayList<>(data);
            
            newData.sort((a, b) -> {
                Object valA = a.get(columnName);
                Object valB = b.get(columnName);
                
                if (valA == null && valB == null) return 0;
                if (valA == null) return ascending ? -1 : 1;
                if (valB == null) return ascending ? 1 : -1;
                
                String type = columnTypes.getOrDefault(columnName, "string");
                int result = 0;
                
                switch (type) {
                    case "number":
                        double numA = valA instanceof Number ? ((Number) valA).doubleValue() : 
                                      Double.parseDouble(valA.toString());
                        double numB = valB instanceof Number ? ((Number) valB).doubleValue() : 
                                      Double.parseDouble(valB.toString());
                        result = Double.compare(numA, numB);
                        break;
                    case "date":
                        // Simple string comparison for dates (assumes ISO format)
                        result = valA.toString().compareTo(valB.toString());
                        break;
                    default:
                        result = valA.toString().compareTo(valB.toString());
                }
                
                return ascending ? result : -result;
            });
            
            dataState.set(newData);
        }
        
        return this;
    }
    
    public DataTable filter(String columnName, String value) {
        if (value == null || value.trim().isEmpty()) {
            filters.remove(columnName);
        } else {
            filters.put(columnName, value.toLowerCase());
        }
        
        applyFilters();
        return this;
    }
    
    public DataTable clearFilters() {
        filters.clear();
        dataState.set(data);
        return this;
    }
    
    private void applyFilters() {
        if (filters.isEmpty()) {
            return;
        }
        
        List<Map<String, Object>> filteredData = new ArrayList<>();
        
        for (Map<String, Object> row : data) {
            boolean matches = true;
            
            for (Map.Entry<String, String> filter : filters.entrySet()) {
                String columnName = filter.getKey();
                String filterValue = filter.getValue();
                
                if (!row.containsKey(columnName) || row.get(columnName) == null) {
                    matches = false;
                    break;
                }
                
                String cellValue = row.get(columnName).toString().toLowerCase();
                if (!cellValue.contains(filterValue)) {
                    matches = false;
                    break;
                }
            }
            
            if (matches) {
                filteredData.add(row);
            }
        }
        
        dataState.set(filteredData);
    }
    
    public DataTable loadFromApi(String url) {
        XMLHttpRequest xhr = XMLHttpRequest.create();
        xhr.open("GET", url);
        xhr.setOnReadyStateChange(() -> {
            if (xhr.getReadyState() == XMLHttpRequest.DONE) {
                if (xhr.getStatus() == 200) {
                    try {
                        JSObject jsonData = JSON.parse(xhr.getResponseText());
                        List<Map<String, Object>> newData = convertJSArrayToList(jsonData);
                        dataState.set(newData);
                    } catch (Exception e) {
                        System.err.println("Error parsing JSON: " + e.getMessage());
                    }
                }
            }
        });
        xhr.send();
        return this;
    }
    
    @SuppressWarnings("unchecked")
    private List<Map<String, Object>> convertJSArrayToList(JSObject jsArray) {
        List<Map<String, Object>> result = new ArrayList<>();
        int length = getArrayLength(jsArray);
        
        for (int i = 0; i < length; i++) {
            JSObject jsObject = getArrayItem(jsArray, i);
            Map<String, Object> row = new HashMap<>();
            
            for (String key : getObjectKeys(jsObject)) {
                Object value = getObjectProperty(jsObject, key);
                row.put(key, value);
            }
            
            result.add(row);
        }
        
        return result;
    }
    
    @JSBody(params = {"array"}, script = "return array.length;")
    private static native int getArrayLength(JSObject array);
    
    @JSBody(params = {"array", "index"}, script = "return array[index];")
    private static native JSObject getArrayItem(JSObject array, int index);
    
    @JSBody(params = {"object"}, script = "return Object.keys(object);")
    private static native String[] getObjectKeys(JSObject object);
    
    @JSBody(params = {"object", "key"}, script = "return object[key];")
    private static native Object getObjectProperty(JSObject object, String key);
    
    private void refreshTable() {
        if (getElement() == null) {
            return;
        }
        
        getElement().setInnerHTML("");
        
        // Create table element
        HTMLElement table = HTMLDocument.current().createElement("table");
        table.getStyle().setProperty("width", "100%");
        table.getStyle().setProperty("border-collapse", "collapse");
        table.getStyle().setProperty("font-size", "14px");
        
        // Create header
        if (!columnNames.isEmpty()) {
            HTMLElement thead = HTMLDocument.current().createElement("thead");
            HTMLElement headerRow = HTMLDocument.current().createElement("tr");
            
            for (String columnName : columnNames) {
                HTMLElement th = HTMLDocument.current().createElement("th");
                th.setTextContent(columnName);
                th.getStyle().setProperty("padding", "12px 10px");
                th.getStyle().setProperty("text-align", "left");
                th.getStyle().setProperty("border-bottom", "2px solid #ddd");
                th.getStyle().setProperty("font-weight", "600");
                th.getStyle().setProperty("background-color", "#f5f5f5");
                th.getStyle().setProperty("cursor", "pointer");
                
                // Add sort indicator if this column is sorted
                if (columnName.equals(sortColumn)) {
                    th.setTextContent(columnName + (sortAscending ? " ▲" : " ▼"));
                }
                
                // Add sort functionality
                final String colName = columnName;
                th.addEventListener("click", e -> {
                    boolean asc = !colName.equals(sortColumn) || !sortAscending;
                    sortBy(colName, asc);
                });
                
                headerRow.appendChild(th);
            }
            
            thead.appendChild(headerRow);
            table.appendChild(thead);
        }
        
        // Create filter row
        if (!columnNames.isEmpty() && !filters.isEmpty()) {
            HTMLElement filterRow = HTMLDocument.current().createElement("tr");
            filterRow.getStyle().setProperty("background-color", "#f9f9f9");
            
            for (String columnName : columnNames) {
                HTMLElement td = HTMLDocument.current().createElement("td");
                td.getStyle().setProperty("padding", "8px 10px");
                
                if (filters.containsKey(columnName)) {
                    HTMLElement input = HTMLDocument.current().createElement("input");
                    input.setAttribute("type", "text");
                    input.setAttribute("placeholder", "Filter...");
                    input.setAttribute("value", filters.get(columnName));
                    input.getStyle().setProperty("width", "100%");
                    input.getStyle().setProperty("padding", "4px");
                    input.getStyle().setProperty("border", "1px solid #ddd");
                    input.getStyle().setProperty("border-radius", "3px");
                    
                    final String colName = columnName;
                    input.addEventListener("input", e -> {
                        String value = input.getAttribute("value");
                        filter(colName, value);
                    });
                    
                    td.appendChild(input);
                }
                
                filterRow.appendChild(td);
            }
            
            table.appendChild(filterRow);
        }
        
        // Create body
        HTMLElement tbody = HTMLDocument.current().createElement("tbody");
        
        for (int i = 0; i < data.size(); i++) {
            final int rowIndex = i;
            Map<String, Object> rowData = data.get(i);
            HTMLElement row = HTMLDocument.current().createElement("tr");
            
            if (selectable) {
                row.getStyle().setProperty("cursor", "pointer");
                
                row.addEventListener("click", e -> {
                    selectedRow = rowIndex;
                    updateSelection();
                    
                    if (selectionListener != null) {
                        selectionListener.accept(rowIndex);
                    }
                    
                    if (rowClickListener != null) {
                        rowClickListener.accept(rowData);
                    }
                });
                
                if (rowIndex == selectedRow) {
                    row.getStyle().setProperty("background-color", "#e8f0fe");
                }
            }
            
            // Add hover effect
            row.addEventListener("mouseover", e -> {
                if (rowIndex != selectedRow) {
                    row.getStyle().setProperty("background-color", "#f5f5f5");
                }
            });
            
            row.addEventListener("mouseout", e -> {
                if (rowIndex != selectedRow) {
                    row.getStyle().setProperty("background-color", "");
                }
            });
            
            // Add cells
            for (String columnName : columnNames) {
                HTMLElement td = HTMLDocument.current().createElement("td");
                td.getStyle().setProperty("padding", "10px");
                td.getStyle().setProperty("border-bottom", "1px solid #ddd");
                
                Object cellValue = rowData.get(columnName);
                td.setTextContent(cellValue != null ? cellValue.toString() : "");
                
                row.appendChild(td);
            }
            
            tbody.appendChild(row);
        }
        
        table.appendChild(tbody);
        getElement().appendChild(table);
        
        // Add empty state message if no data
        if (data.isEmpty()) {
            HTMLElement emptyState = HTMLDocument.current().createElement("div");
            emptyState.setTextContent("No data available");
            emptyState.getStyle().setProperty("padding", "20px");
            emptyState.getStyle().setProperty("text-align", "center");
            emptyState.getStyle().setProperty("color", "#888");
            
            getElement().appendChild(emptyState);
        }
    }
    
    private void updateSelection() {
        if (getElement() == null) {
            return;
        }
        
        HTMLElement table = (HTMLElement) getElement().querySelector("table");
        if (table == null) return;
        
        HTMLElement tbody = (HTMLElement) table.querySelector("tbody");
        if (tbody == null) return;
        
        for (int i = 0; i < tbody.getChildNodes().getLength(); i++) {
            HTMLElement row = (HTMLElement) tbody.getChildNodes().item(i);
            
            if (i == selectedRow) {
                row.getStyle().setProperty("background-color", "#e8f0fe");
            } else {
                row.getStyle().setProperty("background-color", "");
            }
        }
    }
    
    @Override
    public HTMLElement render(HTMLElement parent) {
        HTMLElement element = super.render(parent);
        refreshTable();
        return element;
    }
    
    public List<Map<String, Object>> getData() {
        return new ArrayList<>(data);
    }
    
    public State<List<Map<String, Object>>> getDataState() {
        return dataState;
    }
}